components/compass-runtime-agent/internal/kyma/secrets/credentialsservice.go

```diff
@@ -0,0 +1,110 @@
+	"github.com/kyma-project/kyma/components/compass-runtime-agent/internal/kyma/applications"
+	"github.com/kyma-project/kyma/components/compass-runtime-agent/internal/kyma/model"
+	"github.com/kyma-project/kyma/components/compass-runtime-agent/internal/kyma/secrets/strategy"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+type modificationFunction func(modStrategy strategy.ModificationStrategy, application string, appUID types.UID, name, packageID string, newData strategy.SecretData) apperrors.AppError
+
+//go:generate mockery --name CredentialsService
+type CredentialsService interface {
+	Get(application string, credentials applications.Credentials) (model.Credentials, apperrors.AppError)
+	Create(application string, appUID types.UID, packageID string, credentials *model.Credentials) (applications.Credentials, apperrors.AppError)
+	Upsert(application string, appUID types.UID, packageID string, credentials *model.Credentials) (applications.Credentials, apperrors.AppError)
+	Delete(name string) apperrors.AppError
+}
+
+type credentialsService struct {
+	repository      Repository
+	strategyFactory strategy.Factory
+	nameResolver    k8sconsts.NameResolver
+}
+
+func NewCredentialsService(repository Repository, strategyFactory strategy.Factory, nameResolver k8sconsts.NameResolver) CredentialsService {
+	return &credentialsService{
+		repository:      repository,
+		strategyFactory: strategyFactory,
+		nameResolver:    nameResolver,
+	}
+}
+
+func (s *credentialsService) Create(application string, appUID types.UID, packageID string, credentials *model.Credentials) (applications.Credentials, apperrors.AppError) {
+	return s.modifySecret(application, appUID, packageID, credentials, s.createSecret)
+}
+
+func (s *credentialsService) Get(application string, credentials applications.Credentials) (model.Credentials, apperrors.AppError) {
+	accessStrategy, err := s.strategyFactory.NewSecretAccessStrategy(&credentials)
+	if err != nil {
+		return model.Credentials{}, err.Append("Failed to initialize strategy")
+	}
+
+	data, err := s.repository.Get(credentials.SecretName)
+	if err != nil {
+		return model.Credentials{}, err
+	}
+
+	return accessStrategy.ToCredentials(data, &credentials)
+}
+
+func (s *credentialsService) Upsert(application string, appUID types.UID, packageID string, credentials *model.Credentials) (applications.Credentials, apperrors.AppError) {
+	return s.modifySecret(application, appUID, packageID, credentials, s.upsertSecret)
+}
+
+func (s *credentialsService) Delete(name string) apperrors.AppError {
+	return s.repository.Delete(name)
+}
+
+func (s *credentialsService) modifySecret(application string, appUID types.UID, packageID string, credentials *model.Credentials, modFunction modificationFunction) (applications.Credentials, apperrors.AppError) {
+	if credentials == nil {
+		return applications.Credentials{}, nil
+	}
+
+	modStrategy, err := s.strategyFactory.NewSecretModificationStrategy(credentials)
```

Comment:
I registered the following service in Application Connectivity Adapter:
```{

"provider": "SAP",
"name": "test-proxy-no-auth",
"description": "httpbin.org",
"api": {
  "targetUrl": "https://httpbin.org"
}
}
```
I noticed the following error in the Compass Runtime Agent log:
```
time="2021-05-26T10:42:21Z" level=warning msg="Failed to update credentials secrets for application 'mp-cr-gw-1': , , Failed to initialize strategy, Error: only one credential type have to be provided.{%!!(int=4) %!!(string=Failed to initialize strategy, Error: only one credential type have to be provided.)}(EXTRA apperrors.appError=Failed to initialize strategy, Error: only one credential type have to be provided.)."
```

Reply:
Fixed.